### PR TITLE
Range, then bucket

### PR DESCRIPTION
**Pull request by cavis**
Okay, after several re-refactors ... here we go!  See #47.

Internal changes:

- The `%BigQuery.Interval{}` struct now also has a `bucket` key
  - The old `intv.rollup` will be used in querying bigquery, and `intv.bucket` will be used to format the data before it's returned.  (Example: "rollup" by day, but "bucket" by month).
- Data is only queried/cached by 1h or 1d intervals.
- Had to do a few hacks to keep `?group=` queries working.  Since they aren't cached the same way as this other data.  It's still a TODO to figure out how to better cache that data.
- Many test changes to support removing the 15m interval.

API changes:

- The `?to` date is now truly exclusive.  So if you ask for `?from=2018-02-10&to=2018-02-19` ... the last bucket of data returned is `2018-02-18`.
- The returned json `{interval: "DAY"}` is now more specific.  Something like `{interval: {from: "", to: "", name: "DAY"}}`.  Makes things easier to debug.
- 15m interval gone

### CODE REVIEW

File: lib/big_query/base/timestamp.ex [hunk 1]
```diff
@@ -30,7 +30,7 @@ defmodule BigQuery.Base.Timestamp do
 
   def timestamp_params(interval) do
     lower = interval.rollup.floor(interval.from)
-    upper = interval.rollup.next(interval.to)
+    upper = interval.rollup.ceiling(interval.to)
```
Comment (cavis) [inline lib/big_query/base/timestamp.ex]: (upper `?to=` timestamp is now _exclusive_)

File: lib/big_query/timestamp_rollups/daily.ex [hunk 2]
```diff
@@ -24,10 +24,10 @@ defmodule BigQuery.TimestampRollups.Daily do
   end
 
   def range(from, to) do
-    range(floor(from), floor(to), [])
+    range(floor(from), ceiling(to), [])
```
Comment (cavis) [inline lib/big_query/timestamp_rollups/daily.ex]: All of these rollups changed somewhat ... mostly to standardize how they did ranges.  Check the tests to actually see how they behave (the range array no longer includes the `?to=` date).

File: lib/castle_web/controllers/api/download_controller.ex [hunk 3]
```diff
@@ -10,35 +10,53 @@ defmodule CastleWeb.API.DownloadController do
...
+  defp cache_episode(intv, guid) do
+    @redis.interval "downloads.episodes", intv, guid, fn(new_intv) ->
       @bigquery.episode_downloads(new_intv)
     end
-    render conn, IntervalView, "episode.json", guid: guid, interval: intv.rollup.name,
-      downloads: data, meta: meta
+  end
+
+  defp group_cache_podcast(intv, group, id) do
+    intv_nobucket = intv |> Map.put(:rollup, intv.bucket) |> Map.put(:bucket, nil)
```
Comment (cavis) [inline lib/castle_web/controllers/api/download_controller.ex]: This is a hack!  The `intv.rollup` is now hourly/daily, whereas `intv.bucket` is what you actually specified in `?interval=1d`.  So for the somewhat deprecated `?group=` queries, I just copy the bucket -> rollup.

File: lib/plugs/interval/rollups.ex [hunk 4]
```diff
@@ -1,42 +1,16 @@
...
-
-  defp validate_window(%{assigns: %{interval: %{from: from, to: to}}}, rollup) do
-    if rollup.count_range(from, to) > @max_in_window do
-      {:error, "Time window too large for specified interval"}
-    else
-      {:ok, rollup}
+  def parse(%{assigns: %{interval: %{bucket: bucket}}}) do
+    case bucket.name() do
+      "HOUR" -> {:ok, @hourly}
+      "DAY" -> {:ok, @hourly}
```
Comment (gcampo88) [inline lib/plugs/interval/rollups.ex]: Is this hourly so we can see partial data for today?

File: test/controllers/api/bucket_helper_test.exs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  test "handles data after last bucket", %{buckets: buckets} do
+    raw = [
+      %{time: get_dtim("2017-03-27T00:00:00Z"), count: 4},
+      %{time: get_dtim("2018-01-01T00:00:01Z"), count: 5},
+    ]
+    data = bucketize(raw, buckets)
+    assert_times(data)
+    assert Enum.at(data, 0).count == 0
+    assert Enum.at(data, 1).count == 4
+    assert Enum.at(data, 2).count == 5
```
Comment (gcampo88) [inline test/controllers/api/bucket_helper_test.exs]: so for this one, looks like the 5 downloads from 8 months after the last bucket are being counted in the last `day` bucket? is that right?

File: lib/plugs/interval/rollups.ex (hunk 4 above)
Comment (cavis) [reply lib/plugs/interval/rollups.ex]: Yeah, I think that's what we agreed to in the initial ticket/discussion.  You can ask for partial-days (rounded to the hour), or partial weeks/months (rounded to the day).

File: test/controllers/api/bucket_helper_test.exs (hunk 5 above)
Comment (cavis) [reply test/controllers/api/bucket_helper_test.exs]: Yep.  In practice, this helper should never get called with out-of-range data.  But wanted to make sure it handled it nonetheless.

**Pull request merged**